File: rating.py
THRESHOLD = 3.5
VOTE_BOUNDS = (1, 5)
K = 0.5


def robust_mean(values: list[float], k: float = K, iters: int = 5) -> float:
    if not values:
        return 0.0

    current = sum(values) / len(values)

    for _ in range(iters):
        sum_w = 0.0
        sum_wx = 0.0

        for x in values:
            d = abs(x - current)
            w = 1.0 / (1.0 + (d / k) ** 2)
            sum_w += w
            sum_wx += w * x

        if sum_w > 0:
            current = sum_wx / sum_w

    return current


def get_rating_status(scores: list[int], total_admins: int):
    """
    Рассчитывает текущую оценку и прогноз на основе списка оценок
    и общего количества админов.
    """
    voted_count = len(scores)
    remaining_count = max(0, total_admins - voted_count)

    estimation = robust_mean(scores)

    min_values = scores + [VOTE_BOUNDS[0]] * remaining_count
    min_estimation = robust_mean(min_values)

    max_values = scores + [VOTE_BOUNDS[1]] * remaining_count
    max_estimation = robust_mean(max_values)

    return {
        "estimation": estimation,
        "min_estimation": min_estimation,
        "max_estimation": max_estimation,
        "is_final": remaining_count <= 0,
        "is_guaranteed_approved": min_estimation >= THRESHOLD,
        "is_guaranteed_rejected": max_estimation < THRESHOLD
    }

File: cf/src/rating.ts
export const THRESHOLD = 3.5
export const VOTE_BOUNDS: readonly [number, number] = [1, 5]
const K = 0.5

export function robustMean(values: number[], k: number = K, iters = 5): number {
    if (values.length === 0) return 0

    let current = values.reduce((a, b) => a + b, 0) / values.length

    for (let i = 0; i < iters; i++) {
        let sumW = 0
        let sumWx = 0
        for (const x of values) {
            const d = Math.abs(x - current)
            const w = 1 / (1 + (d / k) ** 2)
            sumW += w
            sumWx += w * x
        }
        if (sumW > 0) current = sumWx / sumW
    }

    return current
}

export interface RatingStatus {
    estimation: number
    min_estimation: number
    max_estimation: number
    is_final: boolean
    is_guaranteed_approved: boolean
    is_guaranteed_rejected: boolean
}

export function getRatingStatus(scores: number[], totalAdmins: number): RatingStatus {
    const votedCount = scores.length
    const remainingCount = Math.max(0, totalAdmins - votedCount)

    const estimation = robustMean(scores)

    const minValues = scores.concat(Array(remainingCount).fill(VOTE_BOUNDS[0]))
    const minEstimation = robustMean(minValues)

    const maxValues = scores.concat(Array(remainingCount).fill(VOTE_BOUNDS[1]))
    const maxEstimation = robustMean(maxValues)

    return {
        estimation,
        min_estimation: minEstimation,
        max_estimation: maxEstimation,
        is_final: remainingCount <= 0,
        is_guaranteed_approved: minEstimation >= THRESHOLD,
        is_guaranteed_rejected: maxEstimation < THRESHOLD,
    }
}
